runtime/scripts/results.js

```diff
@@ -0,0 +1,211 @@
+                        expectedAnswers.push(gapObject.marks);
+                        studentAnswers.push(gapObject.score);
+                        header.push(questionName + " " + partName + " " + gapName + " Answer");
+                        trueOrder.push(trueQuestionNumber + " " + partName + " " + gapName + " Answer");
+                        expectedAnswers.push(gapObject.getCorrectAnswer(gapObject.getScope()));
+                        studentAnswers.push(gapObject.studentAnswer);
+                    }
+                    for (let stepKey of Object.keys(partObject.steps)) {
+                        let stepObject = partObject.steps[stepKey];  
+                        let stepName = stepObject.name||("Bit "+stepKey);    
+                        header.push(questionName + " " + partName + " " + stepName + " Marks");
+                        trueOrder.push(trueQuestionNumber + " " + partName + " " + stepName + " Marks");
+                        expectedAnswers.push(stepObject.marks);
+                        studentAnswers.push(stepObject.score);
+                        header.push(questionName + " " + partName + " " + stepName + " Answer");
+                        trueOrder.push(trueQuestionNumber + " " + partName + " " + stepName + " Answer");
+                        expectedAnswers.push(stepObject.getCorrectAnswer(stepObject.getScope()));
+                        studentAnswers.push(stepObject.studentAnswer);
+                    }
+                }
+            }
+            let dataset = [header, expectedAnswers, studentAnswers];
+            if (randomised) {
+                dataset = [header, trueOrder, expectedAnswers, studentAnswers];
+            }
+            return Numbas.results.make_csv(dataset);
+        },
+
+        /**
+         * 
+         * @param {string} file 
+         */
+        create_and_download_file : function(file) {
+            //pulled from https://stackoverflow.com/questions/8310657/how-to-create-a-dynamic-file-link-for-download-in-javascript
+            let mime_type = 'text/plain';
+            var blob = new Blob([file],{type: mime_type});
+            var dlink = document.createElement('a');
+            document.body.appendChild(dlink); //may be necessary for firefox/some browsers
+            dlink.download="results.csv";
+            dlink.href = window.URL.createObjectURL(blob);
+            dlink.onclick = function(e) {
+              var that = this;
+              setTimeout(function() {
+                window.URL.revokeObjectURL(that.href);
+              }, 1500);
+            };
+          
+            dlink.click()
+            dlink.remove()
+            
+          },
+        /**
+         * Checks if any aspect of the exam is in a randomised order
+         * 
+         * @returns {boolean} whether the exam has any randomised components
+         */
+        examIsRandomised : function() {

```

Comment:
I don't think it's a good idea for the format of the exam to depend on whether anything is out of the initial order.
For a randomised exam, a small proportion of attempts will have the questions in the original order and the format of the CSV will be different to the others!

Reply:
'isRandomised' checks if the exam has tags relating to randomisation, not if the order is direct - if they randomly got the same order as the original group/question order, it would still display that mapping. At least, unless I've written this very differently to how I intended!

Reply:
To be honest, I didn't even look at the code! But I don't think there's any point in changing the format for exams with random question order.

Reply:
Hm, so we'd just always have the 'true question order' row? The check for randomisation was to avoid the inclusion of unnecessary information, but as I still generate the row (I didn't want to put in a bunch of 'if' in every loop) it may not be an issue, and if I reformat it to be in the q0p0g1a format it should provide the unique headings we need...

I do think that 'isRandomised' (checking if any component is randomised (or potentially randomised)) is a likely useful function, but perhaps not needed here!